File: ch6/clip4/boj14597/main.py
h, w = map(int, input().split())
image = [[[-1] * (w + 2) for _ in range(h + 2)] for _ in range(2)]
dp = [[float('inf')] * (w + 2) for _ in range(h + 2)]

for i in range(w + 2):
    dp[0][i] = 0

for img in range(2):
    for i in range(1, h + 1):
        row = list(map(int, input().split()))
        for j in range(1, w + 1):
            image[img][i][j] = row[j - 1]

for i in range(1, h + 1):
    for j in range(1, w + 1):
        dp[i][j] = min(dp[i - 1][j - 1], dp[i - 1][j], dp[i - 1][j + 1]) + (image[0][i][j] - image[1][i][j]) ** 2

print(min(dp[h][1:w + 1]))

File: ch6/clip4/boj14597/Main.java
import java.util.Arrays;
import java.util.Scanner;


public class Main {
    static int[][] dp;
    static int[][][] image;

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        int h = sc.nextInt();
        int w = sc.nextInt();
        final int MAX_VALUE = 1000000000;
        image = new int[2][h + 2][w + 2];
        dp = new int[h + 2][w + 2];

        for (int img = 0; img <= 1; img++) {
            for (int i = 1; i <= h; i++) {
                for (int j = 1; j <= w; j++) {
                    image[img][i][j] = sc.nextInt();
                }
            }
        }
        for(int i = 1; i <= h; i++) {
            Arrays.fill(dp[i], MAX_VALUE);
        }
        int answer = MAX_VALUE;
        for(int r = 1; r <= h; r++) {
            for(int c = 1; c <= w; c++) {
                int case1 = dp[r - 1][c - 1];
                int case2 = dp[r - 1][c];
                int case3 = dp[r - 1][c + 1];
                int diff = image[0][r][c] - image[1][r][c];
                dp[r][c] = Math.min(case1, Math.min(case2, case3)) +
                    (diff * diff);
            }
        }
        for(int c = 1; c <= w; c++) {
            answer = Math.min(answer, dp[h][c]);
        }
        System.out.println(answer);
    }
}
